tracer/src/Datadog.Trace/ClrProfiler/NativeCallTargetUnmanagedMemoryHelper.cs

```diff
@@ -0,0 +1,403 @@
+﻿// <copyright file="NativeCallTargetUnmanagedMemoryHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Datadog.Trace.ClrProfiler;
+
+internal static class NativeCallTargetUnmanagedMemoryHelper
+{
+    private const int SizeOfMemorySegment = 750 * 1024; // 750Kb
+    private static readonly int SizeOfPointer = Marshal.SizeOf(typeof(IntPtr));
+    private static readonly List<IntPtr> Segments;
+    private static IntPtr _currentSegment;
+    private static int _segmentOffset;
+
+    static NativeCallTargetUnmanagedMemoryHelper()
+    {
+        Segments = new(5);
+        CreateSegment();
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static void CreateSegment()
+    {
+        _currentSegment = Marshal.AllocCoTaskMem(SizeOfMemorySegment);
+        _segmentOffset = 0;
+        Segments.Add(_currentSegment);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static IntPtr Allocate(int bytesCount)
+    {
+        var offset = _segmentOffset;
+        if (SizeOfMemorySegment - offset < bytesCount)
+        {
+            CreateSegment();
+        }
+
+        _segmentOffset = offset + bytesCount;
+        return _currentSegment + offset;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static void Free()
+    {
+        for (var i = 0; i < Segments.Count; i++)
+        {
+            Marshal.FreeCoTaskMem(Segments[i]);
+        }
+
+        _currentSegment = IntPtr.Zero;
+        _segmentOffset = 0;
+        Segments.Clear();
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static unsafe IntPtr AllocateAndWriteUtf16String(string value)
+    {
+        var stringPtrSize = value.Length * 2;
```

Comment:
For `AllocateAndWriteUtf16StringArray` you have a special case when the array is null, maybe here you should have one if the string is null

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/4326/commits/74dbe71587ecf6074f3559b5ccad87ee22ec4cfb